library/std/src/sys/pal/unix/net.rs

```diff
@@ -213,15 +213,9 @@ impl Socket {
                 }
                 0 => {}
                 _ => {
-                    // linux returns POLLOUT|POLLERR|POLLHUP for refused connections (!), so look
-                    // for POLLHUP rather than read readiness
-                    if pollfd.revents & libc::POLLHUP != 0 {
-                        let e = self.take_error()?.unwrap_or_else(|| {
-                            io::const_io_error!(
-                                io::ErrorKind::Uncategorized,
-                                "no error set after POLLHUP",
-                            )
-                        });
+                    // Check if the connnection actually succeeded and return ok only when
+                    // the socket is ready and no errors were found
+                    if let Some(e) = self.take_error()? {
                         return Err(e);
                     }
```

Comment:
I don't think it makes sense to change this for all platforms, since https://github.com/rust-lang/rust/pull/127272 (precursor) and https://github.com/rust-lang/rust/issues/127018 seem to indicate this is only on vxworks (or are there others you know of?)

```suggestion
                    if cfg!(target_os = "vxworks") {
                        // [comment about vxworks not using POLLHUP]
                        if let Some(e) = self.take_error()? {
                            return Err(e);
                        }
                    } else {
                        /* original impl */
                    }
```

Also just for reference, if you put `Fixes #127018` in a line of your PR description then it will close the issue when this merges.

Reply:
@tgross35 I don't think that changes I made will impact other platforms. For code simplicity, I did not make these changes specific for Vxworks.

Reply:
I also do not understand the details here but to me it seems like the difference between returning `Err` on all errors vs. only on hangups must be intentional. So I am hesitant to change this without having a better picture of what is going on.

@sfackler it looks like you authored this bit of code in https://github.com/rust-lang/rust/pull/43062 and updated it in https://github.com/rust-lang/rust/pull/45269. What is the reasoning to check for `POLLOUT`/`POLLHUP` rather than just always returning an error?

Reply:
@the8472 might have a better idea here too

Reply:
I wrote that code 7 years ago - I do not remember the specific context around it.

Reply:
Worth a shot, thanks for taking a look.

Reply:
Thanks @sfackler

Reply:
A few relevant things from https://man7.org/linux/man-pages/man2/poll.2.html:

```text
The field events is an input parameter ...
This field may be specified as zero, in which case the only
events that can be returned in revents are POLLHUP, POLLERR, and
POLLNVAL (see below).
```

We probably don't need to set `POLLOUT` anymore as we aren't checking on it since https://github.com/rust-lang/rust/pull/45269.

```text
       POLLERR
              Error condition (only returned in revents; ignored in
              events).  This bit is also set for a file descriptor
              referring to the write end of a pipe when the read end has
              been closed.

       POLLHUP
              Hang up (only returned in revents; ignored in events).
              Note that when reading from a channel such as a pipe or a
              stream socket, this event merely indicates that the peer
              closed its end of the channel.  Subsequent reads from the
              channel will return 0 (end of file) only after all
              outstanding data in the channel has been consumed.

       POLLNVAL
              Invalid request: fd not open (only returned in revents;
              ignored in events).
```

From https://github.com/rust-lang/rust/issues/45265#issuecomment-336536875, it sounds like in reality it sets all three.

The manpage doesn't say anything about setting an error to `SO_ERROR`. I would assume this is always true, but it does seem safer to me to have the fallback. It seems a bit weird to me that we are checking only against POLLHUP and not a `POLLHUP | POLLNVAL | POLLERR` mask since these are all documented error conditions, even though this does seem to work (perhaps the other cases come up more for files).

---

So: I wouldn't really want to change the existing Linux handling to the proposal (except for maybe checking `POLLNVAL` and `POLLERR` too)  because it means going from documented behavior to empirical behavior, even if that seems correct. In which case I would prefer the `cfg` suggestion above to the current proposal.

However, I think this code could be cleaned up a bit with the following, that should work on your system too:

1. Change the `libc::pollfd` struct to have events as 0 rather than `POLLOUT` as we are no longer checking `POLLOUT` (since https://github.com/rust-lang/rust/pull/45269)
2. Since we are no longer have a possible `POLLOUT` `revent`, the possible results of `libc::poll` are:
    - -1: error via errno
    - 0: success
    - 3 (Linux), something else (vxworks): emitted three revents, `POLLHUP | POLLNVAL | POLLERR`
3. Since we no longer have the case of one successful revent (`POLLOUT`), the last match arm can be changed to always emit an error without maybe checking `revents`. So just remove the `if` from the existing code so it always emits `self.take_error()?.unwrap_or_else(|| { ... })`.

   That is, it will always try to read the `SO_ERROR` (works on both Linux and vxworks). If that fails, meaning revents were set but there is no `SO_ERROR` for some unexpected reason, it will emit an `Uncategorized` error (can never happen on vxworks per https://github.com/rust-lang/rust/issues/127018, but maybe possible on linux?).
4. If you feel like it, `debug_assert!` that the revents are what we expect.

If that tests well, then I think it might be a more robust option on all systems. If you make the changes please comment it well, somebody else is going to have to reverse engineer this one day...

---

Doesn't matter if the above works, but does Vxworks set `POLLNVAL` or `POLLERR` rather than a mask including `POLLHUP`?

Reply:
If we set `poll->events` to zero, poll will wait until the specified time until its gets an file descriptor with error event. This behavior is not convenient for successful connections as you would end up waiting till the specified timeout.
I think going with the cfg option for vxworks is better for now.

Reply:
I'm a bit surprised that no other targets reported an error on connect_timeout only checking POLLHUP in revents.

Reply:
Ah right, duh. Thanks for the change and for all the followup here, I left one suggested change to a comment then looks good.